sketch/src/sketch_rnn/model.ts

```diff
@@ -0,0 +1,363 @@
+  }
+
+  dispose() {
+    if (this.rawVars) {
+      for (let i = 0; i < this.rawVars.length; i++) {
+        this.rawVars[i].dispose();
+      }
+      this.rawVars = undefined;
+    }
+    if (this.forgetBias) {
+      this.forgetBias.dispose();
+      this.forgetBias = undefined;
+    }
+    this.initialized = false;
+  }
+
+  /**
+   * sets the internal EXTRA factor of this model (pixel to model space)
+   *
+   * @param scale (the extra scale factor for pixel to model space)
+   *
+   * @returns nothing
+   */
+  set_pixel_factor(scale: number) {
+    // for best effect, set to 1.0 for d3 or paper.js, 2.0 for p5.js
+    this.pixelFactor = scale;
+    this.scaleFactor = this.info.scale_factor / this.pixelFactor;
+  }
+
+  /**
+   * Updates the RNN, returns the next state.
+   *
+   * @param pen [dx, dy, pen_down, pen_up, pen_end].
+   * @param state previous [c, h] state of the LSTM.
+   *
+   * @returns the next [c, h] state of the LSTM`.
+   */
+  update(pen: number[], state: Float32Array[]) {
+    const out = tf.tidy(() => {
+      const numUnits = this.numUnits;
+      const s = this.scaleFactor;
+      const normalizedPen = [pen[0] / s, pen[1] / s, pen[2], pen[3], pen[4]];
+      const x = tf.tensor2d(normalizedPen, [1, 5]);
+      const c = tf.tensor2d(state[0], [1, numUnits]);
+      const h = tf.tensor2d(state[1], [1, numUnits]);
+      const newState = tf.basicLSTMCell(
+        this.forgetBias,
+        this.lstmKernel,
+        this.lstmBias,
+        x,
+        c,
+        h);
+      return newState;
+    });
+    const newC = out[0].dataSync();
+    const newH = out[1].dataSync();
+    for (let i = 0; i < out.length; i++) {
+      out[i].dispose();
+    }
+    return [newC, newH];
+  }
+
+  /**
+   * Given the RNN state, returns the probabilty distribution function (pdf)
+   * of the next stroke. Optionally adjust the temperature of the pdf here.
+   *
+   * @param state previous [c, h] state of the LSTM.
+   * @param temperature (Optional) for dx and dy (default 0.65)
+   * @param softmaxTemperature (Optional) for Pi and Pen discrete states
+   * (default is temperature * 0.5 + 0.5, which is a nice heuristic.)
+   *
+   * @returns [pi, mu1, mu2, sigma1, sigma2, corr, pen]
+   */
+  get_pdf(state: Float32Array[],
+    temperature?: number,
+    softmaxTemperature?: number) {
+    let temp = 0.65;
+    if (temperature) {
+      temp = temperature;
+    }
+    let discreteTemp: number = 0.5 + temp * 0.5; // good heuristic.
+    if (softmaxTemperature) {
+      discreteTemp = softmaxTemperature;
+    }
+    const out = tf.tidy(() => {
+      const numUnits = this.numUnits;
+      const h = tf.tensor2d(state[1], [1, numUnits]);
+      const NOUT = this.N_MIXTURE;
+
+      const sqrttemp = tf.scalar(Math.sqrt(temp));
+      const softtemp = tf.scalar(discreteTemp);
+
+      const z = tf.add(tf.matMul(h, this.outputKernel), this.outputBias)
+        .squeeze();
+      const pen = tf.softmax(z.slice([0], [3]).div(softtemp));
```

Comment:
can you do a split instead of all of these slices? it would be more efficient and readable

Reply:
Okay, let me try- these were converted from the original model directly